architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,641 @@
+
+She can use tools like Bicep, GitOps, or the `rad` CLI to register the new resource types in any tenant of Radius.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+#### Removing default resource types
+
+Suzy is a platform engineer at Contoso working on the internal developer platform. She wants to avoid confusion about the set of resource types supported by the platform.
+
+The application teams are Contoso are allowed to use `Applications.Datastores/redisCaches` and `Applications.Datastores/postgreSQLDatabases`. They can't use other databases like `Applications.Datastores/sqlServerDatabases`. Suzy uses the `rad` CLI to unregister other types in the `Applications.Datastores` namespace.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+## User Experience (if applicable)
+
+See API design. Tooling and user-experiences are out of scope for this document.
+
+## Design
+
+### High Level Design
+
+The registration APIs allow a user to control the set of resource types known to UCP. Since UCP is responsible for routing, it needs to know the set of resource providers and resource types so that it can send them traffic. 
+
+While resource types are the primary concept for **consumers** of extensibility (eg: application developers using Radius), resource providers are a significant concept for users that **implement** extensibility. The API design proposed here includes resource providers as a concept, and supports important scenarios for operating and versioning resource providers over time, based on lessons-learned from ARM.
+
+For anyone skeptical about this, the concept of resource provider is already present in **every** resource id/URL. This is the meaning of the `providers` segment. Currently URLs like `/planes/radius/local/providers` don't have any functionality associated with them, and return a 404 - this proposal addresses this gap.
+
+One of these lessons-learned is the ability to register multiple regions/locations where a resource provider can operate. In particular users will need the ability to gradually roll out functionality across regions rather than updating all regions at once. This is complex, and doesn't help us in the short term. The approach taken here is aligned with our current strategy for region/location in Radius. The APIs are designed for multi-region but the implementation assumes a single region is used.
+
+#### Resource Types
+
+The registration APIs add a few new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+- `System.Resources/resourceProviders/resourceTypes/apiVersions` for API versions of resource types.
+- `System.Resources/resourceProviders/locations` for the configuration of a resource provider in a specific location/region.
```

Comment:
I think about location as an attribute of an environment rather than a resource provider. By tying a location to a resource provider, are we possibly, we are leaking the implementation? e.g., the location needs a controller/microservice to be running in each location. 

I'm seeing it as:
- resourceProvider as a set of globally scoped objects
- location as a set of globally scoped objects
- environments as an instance of a location and a set of available resource providers within that environment

Reply:
I just read the locations section below. I think the discussion is how environments fit into the resource provider (RP) and location configuration, if at all. 

If I wanted to control the roll-out of a RP by location, it seems like I would register a new RP, or a new version of an existing RP. Then update the environment by adding the new RP or incrementing the version of an existing RP.